rust/definitions/src/navigation.rs

```diff
@@ -0,0 +1,659 @@
+    pub method: Option<Vec<TransactionCard>>,
+    pub new_specs: Option<Vec<TransactionCard>>,
+    pub verifier: Option<Vec<TransactionCard>>,
+    pub warning: Option<Vec<TransactionCard>>,
+    pub types_info: Option<Vec<TransactionCard>>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MTransaction {
+    pub content: TransactionCardSet,
+    pub ttype: TransactionType,
+    pub author_info: Option<TransactionAuthor>,
+    pub network_info: Option<TransactionNetworkInfo>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SeedNameCard {
+    pub seed_name: String,
+    pub identicon: String,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MSeeds {
+    pub seed_name_cards: Vec<SeedNameCard>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MKeyDetails {
+    pub qr: String,
+    pub pubkey: String,
+    pub base58: String,
+    pub identicon: String,
+    pub seed_name: String,
+    pub path: String,
+    pub network_title: String,
+    pub network_logo: String,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MNewSeed {
+    pub keyboard: bool,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MRecoverSeedName {
+    pub keyboard: bool,
+    pub seed_name: String,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MRecoverSeedPhrase {
+    pub keyboard: bool,
+    pub seed_name: String,
+    pub user_input: String,
+    pub guess_set: Vec<String>,
+    pub draft: Vec<SeedWord>,
+    pub ready_seed: Option<String>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SeedWord {
+    pub order: u32,
+    pub content: String,
+}
+
+#[derive(Clone, Default, Debug, PartialEq)]
+pub struct DerivationCheck {
+    pub button_good: bool,
+    pub where_to: Option<DerivationDestination>,
+    pub collision: Option<Address>,
+    pub error: Option<String>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Address {
+    pub base58: String,
+    pub path: String,
+    pub has_pwd: bool,
+    pub identicon: String,
+    pub seed_name: String,
+    pub multiselect: Option<bool>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum DerivationDestination {
+    Pwd,
+    Pin,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MDeriveKey {
+    pub seed_name: String,
+    pub network_title: String,
+    pub network_logo: String,
+    pub network_specs_key: String,
+    pub suggested_derivation: String,
+    pub keyboard: bool,
+    pub derivation_check: Option<DerivationCheck>,
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct MVerifierDetails {
+    pub public_key: String,
+    pub identicon: String,
+    pub encryption: String,
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct MVerifier {
+    pub ttype: String,
+    pub details: MVerifierDetails,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MMetadataRecord {
+    pub specs_version: String,
+    pub meta_hash: String,
+    pub meta_id_pic: String,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MNetworkDetails {
+    pub base58prefix: u16,
```

Comment:
Even this should be String on settings screen. None of this is really processed or sent further in UI.

Reply:
The reason is how far do the type guarantees go and I would always rather have the guarantees go further (and in this case over the FFI bridge) rather than not. Any UI framework is expected to be perfectly capable of turning a `u8` or a `u32` into a text or do whatever the UI wants to do about showing the actual data to the user. There may be many examples where supplying the actual types over the bridge and actually showing them to the user in UI will definitely be benifitial.

* a hash. You may be willing to display `H256` the way `Display` implementation for it does, you may want to display it as the `LowerHex` implementation does or even display the first part of the hash depending on the situation in the UI.
* a timestamp. Take a look at how the work is done now with timetamps, how it is displayed in the UI and so on. Right now it is sent as a string which is painful because it actually may be displayed in different formats and values due to being in different timestamps etc. R.n. you have to parse a timestamp string and turn it into a proper java type.
* an image. this is obvious.

I will always say that the guarantee is always better then no guarantee and this is the hill i am willing to die on.

Reply:
I certainly agree with you on image type, of course it should not be passed as string. I agree on timestamp in some cases - if we need any functionality like localization or timezone conversion (which we don't need - at least it is not planned yet). I completely disagree about types only purpose of which is being fed into `Text()` - as having rust control how the final conversion is done is just way more reliable and maintainable than letting UI framework do that; furthermore, we can guarantee that the conversion is then done identically in all platforms and changes propagate to them naturally. If you look at the app as it is now, you can find divergence in some elements already, as our 2 frameworks have barely noticeable disagreements on how to interpret some things here and there; and in addition to that, I have forgotten to correct them synchronously.

One solution to keep some type safety would be to implement conversion to `String` (or, if we are going to that trouble, maybe markdowned AttributedString or something like that to make things prettier?) in rust so that both frameworks can call it similarly; however I doubt it will have good impact on idiomacy. We can try that and then cut the intermediate operation out if it is the only thing done to the value indeed.